### DISCUSSION TITLE
How to upgrade CRDs in strimzi 0.19.0

### DISCUSSION THREAD

**Discussion by emahdij** [strimzi/strimzi-kafka-operator#8511] [General]
HI
We want to upgrade the strimzi operator from 0.19.0 to 0.22.1 (now our Kafka version is 2.5.0). We have a problem with CRDs,
When we want to upgrade by the 'helm upgrade' command we will get this error:
```
Error: UPGRADE FAILED: rendered manifests contain a resource that already exists. Unable to continue with an update: CustomResourceDefinition "kafkas.kafka.strimzi.io" in namespace "" exists and cannot be imported into the current release: invalid ownership metadata; label validation error: missing key "app.kubernetes.io/managed-by": must be set to "Helm"; annotation validation error: missing key "meta.helm.sh/release-name": must be set to "kafka-operator"; annotation validation error: missing key "meta.helm.sh/release-namespace": must be set to "NAMESPACE"
```
Because of that, I removed the CRDs and then applied the helm chart, But my cluster was down and I had downtime. Do you know how I can prevent this situation?

**Comment by scholzj** [comment]
If you remove the CRDs, then your custom resources will be deleted as well and with that also your clusters. So you normally should never do that unless you are ready to wipe everything.

**Reply by emahdij** [reply]
Exactly,
I want to upgrade without downtime, but because of that error, I don't have a choice. 
Is there any way to upgrade? How can I pass this error?

**Reply by scholzj** [reply]
I'm not a Helm user. You can for sure do it with the YAML files. With 0.22.1, you should just apply the installation YAMLs. You also need to make sure you are on a supported Kubernetes version which in this case is 1.21 or older.

**Reply by emahdij** [reply]
The difference between these versions is CRDs in the helm chart. In 0.19.0 the CRDs are in the helm templates but in the versions after that, the CRDs are in a separate directory that you have create yourself manually.
I also try to create [them](https://github.com/strimzi/strimzi-kafka-operator/releases/download/0.22.1/strimzi-crds-0.22.1.yaml) manually. But style I have the same problem and the same error.

**Reply by scholzj** [reply]
TBH, you are talking about versions over 2 years old. So nobody remembers such details. But a quick look tells me that the CRDs are in the `crds` subdirectory already in 0.19: https://github.com/strimzi/strimzi-kafka-operator/tree/0.19.0/helm-charts/helm3/strimzi-kafka-operator/crds

In any case, as far s I know, Helm does never upgrade CRDs. So where they are should not really matter then much. You have to upgrade them manually.

**Reply by emahdij** [reply]
You are right. Unfortunately, we applied with helm2, And in this version, it's not available.  On the other hand, our problem is not CRDs manifest because we can find it from the helm template.  Our problem is when we want to apply CRDs we get this error:
```
Error: UPGRADE FAILED: rendered manifests contain a resource that already exists. Unable to continue with an update: CustomResourceDefinition "kafkas.kafka.strimzi.io" in namespace "" exists and cannot be imported into the current release: invalid ownership metadata; label validation error: missing key "app.kubernetes.io/managed-by": must be set to "Helm"; annotation validation error: missing key "meta.helm.sh/release-name": must be set to "kafka-operator"; annotation validation error: missing key "meta.helm.sh/release-namespace": must be set to "NAMESPACE"
```
